channeldb/mp_payment.go

```diff
@@ -224,10 +224,10 @@ func (m *MPPayment) InFlightHTLCs() []HTLCAttempt {
 
 // GetAttempt returns the specified htlc attempt on the payment.
 func (m *MPPayment) GetAttempt(id uint64) (*HTLCAttempt, error) {
-	for _, htlc := range m.HTLCs {
-		htlc := htlc
+	for i := range m.HTLCs {
```

Comment:
Without this change, one of the unit tests in the `routing` package fails:

```
$ make unit pkg=routing
 Running unit tests.
go test  -tags="dev nolog"  -test.timeout=60m github.com/lightningnetwork/lnd/routing
--- FAIL: TestPaymentControlSubscribeFail (0.07s)
    --- FAIL: TestPaymentControlSubscribeFail/register_attempt,_delete_failed_payments (0.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7abd56]

goroutine 4869 [running]:
testing.tRunner.func1.2({0x87a340, 0xd4cbb0})
        /home/gitpod/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
        /home/gitpod/go/src/testing/testing.go:1392 +0x39f
panic({0x87a340, 0xd4cbb0})
        /home/gitpod/go/src/runtime/panic.go:838 +0x207
github.com/lightningnetwork/lnd/routing.testPaymentControlSubscribeFail(0xc00b0dad00, 0x1, 0x0)
        /workspace/lnd/routing/control_tower_test.go:240 +0x4f6
github.com/lightningnetwork/lnd/routing.TestPaymentControlSubscribeFail.func2(0xc000503520?)
        /workspace/lnd/routing/control_tower_test.go:182 +0x25
testing.tRunner(0xc00b0dad00, 0xa2cbf8)
        /home/gitpod/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
        /home/gitpod/go/src/testing/testing.go:1486 +0x35f
FAIL    github.com/lightningnetwork/lnd/routing 0.346s
```

In the previous version of GetAttempt, `htlc` as defined in the for loop is a copy, *not a reference*. That's important because we want PaymentControl.FailAttempt to return the payment object. If the config is set to keep failed payment attempts, FailAttempt writes the failure info to disk and then re-fetches and returns the payment attempt, which works fine. However, if the config is set to delete failed attempts, it reads the attempt from disk, wipes it from disk, modifies it in memory, and then returns it. The in-memory modification is where this problem arises, since we're calling GetAttempt to get a reference to the htlc and append the failure info. And if you look at where the test is failing:

`if *htlcAttempt.Failure != failInfo {`

We're attempting to dereference the failure, which is nil because we appended it to a copy of the htlc.

Reply:
Just added a test in `channeldb/mp_payment_test.go` (TestGetAttemptReturnsReferenceToPayment) that directly tests this behavior.